### PR TITLE
feat: substratools methods as generic method

### PR DESCRIPTION
**Pull request by ThibaultFy**
Signed-off-by: ThibaultFy <50656860+ThibaultFy@users.noreply.github.com>

## Companion PR

https://github.com/Substra/substra-tools/pull/74

## Summary

Refacto of substratools_methods.py to add flexibility and simplify the code. 
The functions executed by substratools does not have name constraint anymore.

## Notes

Asana card: https://app.asana.com/0/1202209633046022/1203200116896589/f

## Please check if the PR fulfills these requirements

- [x] If the feature has an impact on the user experience, the [changelog](https://github.com/substra/substrafl/blob/main/CHANGELOG.md) has been updated
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] The commit message follows the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification

### CODE REVIEW

**Comment by Owlfred** [comment]
```
Usage: /e2e [options] [help]

/e2e may appear anywhere as long as it is on its own line

Options:
  --refs <value>                                         Extra refs (branch or tag) with format REPO=GIT_REF,REPO=GIT_REF.
  Supported repositories: substra-hlf-k8s, orchestrator, substra-backend, substra-frontend, substra-tools, substrafl, substra, substra-tests, substra-ci.
  Example: /e2e --refs substra-backend=some_branch,orchestrator=some_tag (default: {})
  --tests-to-run, --tests <tests-to-run>                 Comma-separated list of tests to run. Valid options: sdk,substrafl,frontend or NONE. (default: "sdk")
  --benchmarks-to-run, --benchmarks <benchmarks-to-run>  Comma-separated list of workflows tests to run. Valid options: mnist, camelyon or NONE. (default: "NONE")
  --orchestrator-mode, --mode <orchestrator-mode>        Comma-separated list of orchestrator modes to run tests for. Valid options: standalone,distributed (default: "standalone,distributed")
  -h, --help                                             display help for command
```

**Comment by ThibaultFy** [comment]
/e2e --refs substra-tools=feat/user-define-function-name,substrafl=chore/refacto-remote-method --tests substrafl --benchmarks mnist,camelyon --mode standalone

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by ThibaultFy** [comment]
/e2e --refs substra-tools=feat/user-define-function-name,substrafl=chore/refacto-remote-method --tests substrafl --benchmarks mnist,camelyon --mode standalone

File: substrafl/remote/register/register.py [hunk 1]
```diff
@@ -386,7 +381,9 @@ def add_metric(
     _check_metric_function(metric_function=metric_function)
 
     class Metric:
-        def score(self, datasamples, predictions_path):
+        def score(self, datasamples, predictions_path, _skip=True):
```
Comment (guilhem-barthes) [inline substrafl/remote/register/register.py:384]: Compare with the use of `**kwargs`

**Review by guilhem-barthes — APPROVED** [review]
Thanks for the work!

File: substrafl/remote/substratools_methods.py [hunk 2]
```diff
@@ -29,207 +29,137 @@ def __init__(
...
+                models.append(self.load_model(m_path))
+            loaded_inputs["shared_states"] = models
 
-        tools.register(self.aggregate)
+        if InputIdentifiers.datasamples in inputs:
+            loaded_inputs["datasamples"] = inputs[InputIdentifiers.datasamples]
 
+        if InputIdentifiers.shared in inputs:
+            loaded_inputs["shared_state"] = (
+                self.load_model(inputs[InputIdentifiers.shared]) if inputs[InputIdentifiers.shared] else None
```
Comment (guilhem-barthes) [inline substrafl/remote/substratools_methods.py:62]: I am not sure about readability

**Pull request merged**